### PR TITLE
Enable media message forwarding

### PR DESCRIPTION
**Pull request by McLoo**
This is a stew mainly of #2886 with some nice chunks of #2236 

- Tested forwarding to existing as well as new (group-)convos.
- Tested on a 5.0.2 device and KK emulator. Couldn't get a play service enabled GB emu - can't use a new TS installation without play services... :disappointed: 

All media works fine as well as text and text + media.

### CODE REVIEW

**Comment by jeremymasters** [comment]
Does this also fix selection of media messages?  In 2.16.3 I can't long-press the image to select it.  I have to long-press in the open space beside it.

**Comment by 2-4601** [comment]
@jeremymasters See: #2333, #2488, #2648 (PR).

**Comment by rhodey** [comment]
hey @McLoo, I just checked out this PR, rebased it with master, and ran some tests. every time i try to forward an image message the image preview is the red error icon, tapping send results in no image being sent.

can you rebase this and let me know if you can reproduce? I tested this on my Nexus 5 running stock Android 5.1 Lollipop, rebased into f72cd5b0fffe2465b40213828a9214c8c2517715

**Comment by McLoo** [comment]
@rhodey then you collabs broke it :wink: 

Will check out. The magic is in those lines:
```JAVA
if (isAuthorityPartDb(uri))
  part.setDataUri(PartAuthority.getPublicPartUri(uri));
else
  part.setDataUri(uri);
``` 

Had the triangle before, when not using getPublicPartUri?!

**Comment by McLoo** [comment]
@rhodey tested on CM 12.1
Still works for me. Can you post a debug log? :grin:

**Comment by rhodey** [comment]
hey @McLoo, I just rebased this with 2.18.1 and tested on my stock 2.3.6 Gingerbread device & my stock 4.4.2 device, both work great. my nexus 5 is my daily use phone, I keep my actual install of TextSecure on there, so when I want to test a new build of TS on that device I have to alter the application ID and some other package names in the build.grade and AndroidManifest.xml file-- basically the app that I test on this device is not the exact app that I should be testing.

so, in conclusion it probably doesn't work on my nexus 5 for reasons related to me changing the application ID and package names. interestingly nothing was actually caught in the log file related to the error icon showing up. this PR looks good to me and has tested well, if we could get someone to test this on a stock lollipop device I'd be super confident in merging it.

:+1:

**Comment by McLoo** [comment]
I'm thinking of `assertMediaSize` everytime - not only if not `isAuthorityPartDb`.
There could be changes for media size, after the media went into the DB?!

(`assertMediaSize` should probably move into [sendMediaMessage](https://github.com/WhisperSystems/TextSecure/blob/6caee6b0f076a1dd2e710212435cf950352817db/src/org/thoughtcrime/securesms/ConversationActivity.java#L1108), once bigger files are allowed for push)

**Comment by McLoo** [comment]
Finally I found someone who was willing to try this PR on his daily use phone. :grimacing: 
Titanium Backup works great if you have to juggle with builds while keeping data.

So: on my Sony Z1 compact 5.0.2 (rooted) this also works fine. Couldn't test media playback though cause of #3436

File: src/org/thoughtcrime/securesms/mms/PartAuthority.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
   private static final String PART_URI_STRING   = "content://org.thoughtcrime.securesms/part";
   private static final String THUMB_URI_STRING  = "content://org.thoughtcrime.securesms/thumb";
-  private static final Uri    PART_CONTENT_URI  = Uri.parse(PART_URI_STRING);
+  public  static final Uri    PART_CONTENT_URI  = Uri.parse(PART_URI_STRING);
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/mms/PartAuthority.java]: this is intentionally private

File: src/org/thoughtcrime/securesms/mms/AudioSlide.java [hunk 2]
```diff
@@ -56,24 +55,15 @@ public boolean hasAudio() {
...
-    } finally {
-      if (cursor != null)
-        cursor.close();
-    } 
+    if (isAuthorityPartDb(uri)) {
+      part.setContentType(getContentTypeForPartInDb(context, uri));
+    } else {
+      assertMediaSize(context, uri);
+      part.setContentType(getContentTypeForPartOutsideDb(context, uri, Audio.Media.MIME_TYPE));
+    }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/mms/AudioSlide.java]: we shouldn't be distinguishing, a uri is a uri...

Comment (McLoo) [inline src/org/thoughtcrime/securesms/mms/AudioSlide.java]: But PartProvider has 
`CONTENT_URI_STRING = "content://org.thoughtcrime.provider.securesms/part";`
whereas PartAuthority has 
`PART_URI_STRING   = "content://org.thoughtcrime.securesms/part";` 

the latter leads to an error on forwarding
```
org.thoughtcrime.securesms E/ActivityThread﹕ Failed to find provider info for org.thoughtcrime.securesms
```
as there is no such provider in the manifest.

So at some point we have to decide how the uri has to be interpreted. Or am I completely wrong?!
Or does PartAuthority need a change?

Comment (McLoo) [inline src/org/thoughtcrime/securesms/mms/AudioSlide.java]: changing PartAuthority's PART_URI_STRING to "content://org.thoughtcrime.provider.securesms/part";works fine form me

**Comment by McLoo** [comment]
@moxie0 comments addressed

File: src/org/thoughtcrime/securesms/mms/AudioSlide.java (hunk 2 above)
Comment (McLoo) [inline src/org/thoughtcrime/securesms/mms/AudioSlide.java]: Sorry for spamming. The above comments are concerning Images.

And yet we have to distinguish if we don't want to on fail grabbing parts' contentType

**Comment by McLoo** [comment]
rebased

**Comment by McLoo** [comment]
rebased. reduced refactoring.

**Comment by McLoo** [comment]
:angry: temporarily closing til I'm happy with GIF-detection

getting the MimeType isn't very funny for internal media :/
